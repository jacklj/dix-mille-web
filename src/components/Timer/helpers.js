import moment from 'moment';

export const calculateTimerValue = ({
  countDirection,
  timeLeftOrElapsedWhenLastStarted,
  timeStarted,
}) => {
  const now = moment();
  const timeElapsedSinceStarted = now.diff(timeStarted); // in ms
  // number of ms between the timestamp when the timer was just started and now

  let timerValueMs;
  if (countDirection === 'up') {
    timerValueMs = timeLeftOrElapsedWhenLastStarted + timeElapsedSinceStarted;
  } else if (countDirection === 'down') {
    timerValueMs = timeLeftOrElapsedWhenLastStarted - timeElapsedSinceStarted;
    if (timerValueMs < 0) {
      // in case the timer elapsed, and then we opened the operator app some time later
      timerValueMs = 0;
    }
  } else {
    throw new Error('Unrecognised Timer countDirection', countDirection);
  }
  return timerValueMs;
};

const renderNumberWithLeading0 = (number) => {
  let prettyNumber;
  if (!number || number <= 0) {
    prettyNumber = '00';
  } else if (number > 0 && number < 10) {
    prettyNumber = `0${number}`;
  } else {
    prettyNumber = number;
  }
  return prettyNumber;
};

export const renderTime = (timeInMs) => {
  const mins = Math.floor(timeInMs / 60000);
  const prettyMins = renderNumberWithLeading0(mins);
  const secs = Math.floor(timeInMs / 1000) % 60;
  const prettySecs = renderNumberWithLeading0(secs);
  return `${prettyMins}:${prettySecs}`;
};
