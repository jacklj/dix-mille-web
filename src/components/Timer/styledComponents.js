import React from 'react';
import styled, { css, keyframes } from 'styled-components';
// import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
// import { faPause } from '@fortawesome/free-solid-svg-icons';

const opacityPulseAnimation = keyframes`
  0% { opacity: 0.3; }
  50% { opacity: 1; }
  100% { opacity: 0.3; }
`;

export const TimerStyled = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  align-items: flex-start;
  margin: 5px;
  color: white;
  font-size: 2em;
  line-height: 1em;

  animation: ${(props) =>
    props.isPaused
      ? css`
          ${opacityPulseAnimation} 1.5s ease-in-out 0s infinite
        `
      : css`none`};
`;

export const TimeText = styled.div`
  font-family: 'Electrolize', sans-serif;
  flex: 1;
`;

const Icon = styled.div`
  height: 0.9em;
  width: relative;
  margin-right: 10px;
  flex: 0;
`;

export const PauseIcon = <Icon>paused</Icon>;
