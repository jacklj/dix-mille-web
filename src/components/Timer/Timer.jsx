import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { TimerStyled, TimeText, PauseIcon } from './styledComponents';
import { calculateTimerValue, renderTime } from './helpers';
// import Constants from 'services/constants';

// const { PAGE_TYPES } = Constants;

// # Timer notes:
//   - can go both up and down
// # All timer events:
//  -

class Timer extends Component {
  constructor(props) {
    super(props);
    let state;
    // Timer state: { timerValueMs, intervalTimerId }
    const {
      countDirection,
      isPaused,
      timeStarted,
      timeLeftOrElapsedWhenLastStarted,
      // currentPageType,
    } = this.props;

    if (!isPaused && !timeStarted) {
      // timer should be running but no timeStarted timestamp set - this is inconsistent,
      // so we most likely just haven't got the data in redux store yet. Don't do anything.
      state = {
        timerValueMs: undefined,
        intervalTimerId: undefined,
      };
    } else if (
      isPaused &&
      !timeLeftOrElapsedWhenLastStarted &&
      timeLeftOrElapsedWhenLastStarted !== 0
    ) {
      // paused but no timeLeftElapsed defined - another inconsistent state
      state = {
        timerValueMs: undefined,
        intervalTimerId: undefined,
      };
    } else if (
      isPaused &&
      (timeLeftOrElapsedWhenLastStarted ||
        timeLeftOrElapsedWhenLastStarted === 0)
    ) {
      // page loaded and timer was already paused - set timeValueMs to timeLeftOrElapsedWhenLastStarted
      state = {
        timerValueMs: timeLeftOrElapsedWhenLastStarted,
        intervalTimerId: undefined,
      };
    } else if (!isPaused && timeStarted) {
      const timeStartedAsDate = new Date(timeStarted);
      const timerValueMs = calculateTimerValue({
        // initialise timer with correct value
        countDirection,
        timeLeftOrElapsedWhenLastStarted,
        timeStarted: timeStartedAsDate,
      });
      const intervalTimerId = setInterval(this.tick, 100);
      state = {
        timerValueMs,
        intervalTimerId,
      };
    } else {
      throw new Error('Unrecognised timer condition.', this.props);
    }

    this.state = state;
  }

  // componentDidUpdate detects changes in props and state. We need it to detect if
  // isPaused changes (ie the game is paused or unpaused),
  componentDidUpdate(prevProps, prevState) {
    if (!this.props.isPaused && !this.props.timeStarted) {
      // Timer should be running (and not on Start Page) but no timeStarted timestamp set - this
      // is inconsistent, so we most likely just haven't got the data in redux store yet. Don't
      // do anything.
      return;
    }

    if (
      this.props.isPaused &&
      !this.state.intervalTimerId &&
      !prevState.timerValueMs
    ) {
      // page loaded and timer was already paused - set timeValueMs to timeLeftOrElapsedWhenLastStarted
      if (
        this.state.timerValueMs !== this.props.timeLeftOrElapsedWhenLastStarted
      ) {
        this.setState({
          timerValueMs: this.props.timeLeftOrElapsedWhenLastStarted,
        });
      }
      return;
    } else if (!prevProps.isPaused && this.props.isPaused) {
      // either the game just paused, or the game was already paused when the
      // game data was loaded into the redux store
      if (!this.state.intervalTimerId) {
        throw new Error('Game just paused, but no interval timer was running?');
      }
      // if the game just paused, stop the timer
      clearInterval(this.state.intervalTimerId);
      this.setState({ intervalTimerId: undefined });
    } else if (!this.props.isPaused && !this.state.intervalTimerId) {
      // also set the initial timerValueMs value (for the first 100ms before this.tick() has run the first time?)
      const {
        countDirection,
        timeLeftOrElapsedWhenLastStarted,
        timeStarted,
      } = this.props;

      const timeStartedAsDate = new Date(timeStarted);

      this.setState({
        timerValueMs: calculateTimerValue({
          // initialise timer with correct value
          countDirection,
          timeLeftOrElapsedWhenLastStarted,
          timeStarted: timeStartedAsDate,
        }),
        intervalTimerId: setInterval(this.tick, 100),
      });
    }
  }

  componentWillUnmount() {
    clearInterval(this.state.intervalTimerId);
    this.setState({ intervalTimerId: undefined });
  }

  tick = () => {
    const {
      countDirection,
      timeLeftOrElapsedWhenLastStarted,
      timeStarted,
    } = this.props;
    if (!timeStarted) {
      return;
    }

    const timeStartedAsDate = new Date(timeStarted);

    const timerValueMs = calculateTimerValue({
      countDirection,
      timeLeftOrElapsedWhenLastStarted,
      timeStarted: timeStartedAsDate,
    });

    if (timerValueMs < 0) {
      this.props.onFinish();
      clearInterval(this.state.intervalTimerId);
      this.setState({ intervalTimerId: undefined, timerValueMs: 0 });
    } else {
      this.setState({ timerValueMs });
    }
  };

  render() {
    const { isPaused, style } = this.props;

    const prettyTime =
      typeof this.state.timerValueMs === 'number'
        ? renderTime(this.state.timerValueMs)
        : 'loading';
    return (
      <TimerStyled isPaused={isPaused} style={style}>
        {isPaused && PauseIcon}
        <TimeText>{prettyTime}</TimeText>
      </TimerStyled>
    );
  }
}

Timer.defaultProps = {
  onFinish: () => {},
  countDirection: 'down',
  timeStarted: null,
  currentPageType: undefined,
};

Timer.propTypes = {
  timeStarted: PropTypes.number.isRequired,
  timeLeftOrElapsedWhenLastStarted: PropTypes.number.isRequired,
  isPaused: PropTypes.bool.isRequired,
  countDirection: PropTypes.oneOf(['up', 'down']),
  onFinish: PropTypes.func,
  currentPageType: PropTypes.string,
};

export default Timer;
