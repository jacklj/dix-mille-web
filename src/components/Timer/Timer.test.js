import React from 'react';
import renderer from 'react-test-renderer';
import Timer from './index';

it('renders correctly', () => {
  const timeStarted = Date.now();
  const isPaused = false;
  const timeLeftOrElapsedWhenLastStarted = 20000;
  const tree = renderer
    .create(
      <Timer
        timeStarted={timeStarted}
        timeLeftOrElapsedWhenLastStarted={timeLeftOrElapsedWhenLastStarted}
        isPaused={isPaused}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly when paused', () => {
  const timeStarted = Date.now();
  const isPaused = true;
  const timeLeftOrElapsedWhenLastStarted = 20000;
  const tree = renderer
    .create(
      <Timer
        timeStarted={timeStarted}
        timeLeftOrElapsedWhenLastStarted={timeLeftOrElapsedWhenLastStarted}
        isPaused={isPaused}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
