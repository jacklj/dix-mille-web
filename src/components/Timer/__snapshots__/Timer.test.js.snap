// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-AxjAm ilbYDH"
>
  <div
    className="sc-AxirZ eJUInj"
  >
    00:19
  </div>
</div>
`;

exports[`renders correctly when paused 1`] = `
<div
  className="sc-AxjAm bXtLJg"
>
  <div
    className="sc-AxiKw kULUfT"
  >
    paused
  </div>
  <div
    className="sc-AxirZ eJUInj"
  >
    00:20
  </div>
</div>
`;
